pkg/controller/atlasproject/auditing.go

```diff
@@ -3,96 +3,87 @@ package atlasproject
 import (
 	"reflect"
 
-	"go.mongodb.org/atlas/mongodbatlas"
-
-	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/pointer"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/translation/audit"
 	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
 	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
 	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
 )
 
-func ensureAuditing(workflowCtx *workflow.Context, project *akov2.AtlasProject) workflow.Result {
-	result := createOrDeleteAuditing(workflowCtx, project.ID(), project)
-	if !result.IsOk() {
-		workflowCtx.SetConditionFromResult(api.AuditingReadyType, result)
-		return result
-	}
-
-	if isAuditingEmpty(project.Spec.Auditing) {
-		workflowCtx.UnsetCondition(api.AuditingReadyType)
-		return workflow.OK()
-	}
-
-	workflowCtx.SetConditionTrue(api.AuditingReadyType)
-	return workflow.OK()
+type auditController struct {
+	ctx     *workflow.Context
+	project *akov2.AtlasProject
+	service audit.AuditLogService
 }
 
-func createOrDeleteAuditing(ctx *workflow.Context, projectID string, project *akov2.AtlasProject) workflow.Result {
-	atlas, err := fetchAuditing(ctx, projectID)
+// reconcile dispatch state transitions
+func (a *auditController) reconcile() workflow.Result {
+	auditInAtlas, err := a.service.Get(a.ctx.Context, a.project.ID())
 	if err != nil {
-		return workflow.Terminate(workflow.ProjectAuditingReady, err.Error())
+		return a.terminate(workflow.Internal, err)
 	}
 
-	if !auditingInSync(atlas, project.Spec.Auditing) {
-		err := patchAuditing(ctx, projectID, prepareAuditingSpec(project.Spec.Auditing))
-		if err != nil {
-			return workflow.Terminate(workflow.ProjectAuditingReady, err.Error())
-		}
-	}
+	isUnset := a.project.Spec.Auditing == nil
+	auditInAKO := audit.NewAuditConfig(a.project.Spec.Auditing.DeepCopy())
 
-	return workflow.OK()
-}
+	if !reflect.DeepEqual(auditInAKO, auditInAtlas) {
+		return a.configure(auditInAKO, isUnset)
+	}
 
-func prepareAuditingSpec(spec *akov2.Auditing) *mongodbatlas.Auditing {
-	if isAuditingEmpty(spec) {
-		return &mongodbatlas.Auditing{
-			Enabled: pointer.MakePtr(false),
-		}
+	if a.project.Spec.Auditing == nil {
+		return a.unmanage()
```

Comment:
nit: this transition is not handled in unit tests.

Reply:
Added